jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1460,6 +1476,9 @@ public void succeeded()
 
             try (AutoLock ignored = _request._lock.lock())
             {
+                if (lockedCompleteCallback())
+                    return;
+
```

Comment:
Can you explain this change? Looks like you are fixing some unrelated issue?

It seams wrong to take action before checking for pending demand and write calls?  

The boolean on that method also feels the wrong way around.

Reply:
The old code produces an NPE when an application returns `false` from `handle()`, and then mistakenly completes the callback (which, at that point, is already completed by Jetty producing a 404).

BY doing the early check on whether the callback is already completed, we avoid the NPE.